python/lsst/alert/stream/alertConsumer.py

```diff
@@ -48,11 +48,20 @@ class AlertConsumer(object):
     """
 
     def __init__(self, topic, schema_files=None, **kwargs):
-        self.consumer = confluent_kafka.Consumer(**kwargs)
-        self.consumer.subscribe([topic])
+        self.topic = topic
+        self.kafka_kwargs = kwargs
         if schema_files is not None:
             self.alert_schema = avroUtils.combineSchemas(schema_files)
 
+    def __enter__(self):
+        self.consumer = confluent_kafka.Consumer(**self.kafka_kwargs)
```

Comment:
Is it a bad idea to get rid of __enter__ or put this back to init?  Because if I don't and I don't use a consumer with `with` then I get an error from line 75 that self has no attribute consumer.  If I put it back then I can use `with` or not use with.

Reply:
Hmm, good question.

The key point here is not `__enter__`, it's `__exit__`. At least for me, if I let Python exit without calling the `close()` method on my consumer, it segfaults. Presumably that's because the consumer doesn't get cleaned up properly. At a minimum, the segfault is ugly; I guess it's also possible that it could cause alerts to get lost (presumably whatever book-keeping is necessary to close down the stream isn't done properly, but I'm only speculating).

By using a context manager (ie, with `with`, `__enter__`, `__exit__`), you can ensure that when you exit the context then `close()` is properly called.

If you don't do that, you have to call `close()` in some other way. Possibilities are calling it manually, or putting it in the destructor (`__del__()` method) of `AlertConsumer`. The former is asking for trouble (someday, somebody will forget); the latter is generally regarded as a bad idea.

So based on the above logic, my preference would be to _always_ use `with AlertConsumer...`, and not create one any other way, since that's the only way to be sure things will get cleaned up properly.

But! You know this code better than I do, and may have use cases I'm not thinking of.

Reply:
Ok, that is fine.  I'll just change my other consumer in printStream.py to use a with and then merge this, if that's ok, or you can if you want, so your printStream.py on this branch will work also.

Reply:
Great! Go ahead.